### DISCUSSION TITLE
Initializing EditorState from Yjs Document

### DISCUSSION THREAD

**Discussion by Bunkerbewohner** [facebook/lexical#4442] [Q&A] [answered]
Hi!

I know how to generate HTML from the Editor on the server using `createHeadlessEditor` and `JSDOM` as described in the [docs](https://lexical.dev/docs/concepts/serialization#html---lexical). 

But my question is, **how do I manually initialize the EditorState from a Yjs Document**? 

The `LexicalCollaborationPlugin` does it somehow, and in the source I found a method called `syncYjsChangesToLexical`, however, it's not clear to me how to properly use it to achieve my goal in this case.

Context:

I'm working on a collaborative text editor for documentation based on Lexical and HocusPocus (hocuspocus.dev).
On the server-side I'm persisting the Yjs document sent by the LexicalCollaborationPlugin to the database. Users that visit the web app should first get a server-side rendered preview of the editor content until on the client-side Lexical and the LexicalCollaborationPlugin (using `HocuspocusProvider`) are loaded. At least that's the idea.

Saving and loading the Ydoc works fine, and via the LexicalCollaborationPlugin the editor is initialized as expected, showing the most recently persisted state. I'm just missing the part where I generate the HTML on the server from the Ydoc.

**Comment by Bunkerbewohner** [comment, marked as answer]
I've been trying to implement something based on the code I saw in `LexicalCollaborationPlugin`, however, I haven't been succesful yet. This is my function right, that I want to use in NodeJS to generate HTML from a Ydoc:

(EDIT: I found the missing piece, see my reply below. I updated this code accordingly)

```typescript
export const MARKDOWN_NODES = [
  HeadingNode,  // import {HeadingNode, QuoteNode} from "@lexical/rich-text";
  ListNode, // ...
  ListItemNode,
  QuoteNode,
  CodeNode,
  CodeHighlightNode,
  TableNode,
  TableCellNode,
  TableRowNode,
  AutoLinkNode,
  LinkNode,
  HorizontalRuleNode,
];

export function createHtmlFromYdoc(document: Doc): string {
  // $generateHTmlFromNodes can only be used on the server-side when JSDOM is globally set up right now
  // see https://github.com/facebook/lexical/issues/3097
  const restoreDOM = globallyInstallJSDOM();

  // The headless editor is used to generate the HTML
  const editor = createHeadlessEditor({
    namespace: 'docma',
    nodes: MARKDOWN_NODES, // same list of nodes as used in the client-side editor
    onError: (error) => {
      console.error(error);
    }
  });

  // Try to initialize the editor from the ydoc
  const dummyId = "dummy-id";
  const dummyProvider: any = {
    awareness: {
      getLocalState: () => null,
      getStates: () => [],
    }
  }
  const copyTarget = new Y.Doc();
  const copyBinding = createBinding(editor, dummyProvider, dummyId, copyTarget, new Map([[dummyId, copyTarget]]));

  // this syncs yjs changes to the lexical editor
  const onYjsTreeChanges = (events, transaction) => {
    syncYjsChangesToLexical(copyBinding, dummyProvider as any, events, false);
  };
  copyBinding.root.getSharedType().observeDeep(onYjsTreeChanges);

  // copy the original document to the copy to trigger the observer which updates the editor
  const update = Y.encodeStateAsUpdate(document);
  Y.applyUpdate(copyTarget, update);
  
  editor.update(() => {}, { discrete: true }); // EDIT: this was missing before

  let html: string;
  editor.getEditorState().read(() => {
    html = $generateHtmlFromNodes(editor, null);
  });

  restoreDOM();

  console.log("created HTML: ", html);

  return html;
}
```

In the debugger I see that the Ydoc has the expected content (some simple paragraph like "Hello World"), and I also see that `onYjsTreeChanges` is called and calls  `syncYjsChangesToLexical`. However, the result of `$generateHtmlFromNodes` is always an empty string. Looks like the editor state is not correctly initialized from the Ydoc.

Does anyone have an idea what I'm doing wrong? Or if there's a simpler way to achieve this?

**Reply by Bunkerbewohner** [reply]
I finallly found a [snippet](https://gist.github.com/fantactuka/ee8ef8ac1c51d179c7a9199262444b04) by @fantactuka doing something similar, and there I found the missing line in my code: 

```
editor.update(() => {}, { discrete: true });
```

Apparently I needed to call this after `syncYjsChangesToLexical` to actually update the EditorState. 
I don't understand why that's necessary, considering that internally `syncYjsChangesToLexical` already uses `editor.update(...)`.
Anyway, I'm just happy that I finally got this to work, thanks to @fantactuka!

**Reply by rizwansh-kore** [reply]
Hi @Bunkerbewohner, Thanks for sharing the solution. I was looking for the same thing came across your snippet. This works very well. Appreciate your help.. 💯

**Reply by matsuyama-k1** [reply]
Thank you!
Additionally I imported ignore-styles because loading css in CommonJS causes an error.

import 'ignore-styles';

**Comment by kaikun213** [comment]
Hey @Bunkerbewohner, 

Thanks for sharing the solution!
I am currently trying to build something very similar - backend editor init & conversion to markdown.
Starting from your and @fantactuka's snippet. The document syncs and persists correctly, when I load the ydoc from the websocket. However, **initialisation of the editor does result in an empty export**. Specifically, `onYjsTreeChanges` is triggered and the event targets seem to contain the `_collabNode` field but the editor state does not seem to change.

Did you experience a similar issue? Any help would be highly appreciated!

Here a minimal version of your code adjusted:
```typescript
function createMarkdownFromYdoc(yDocState: Uint8Array, nodes: Array<Klass<LexicalNode>> = MARKDOWN_NODES): string {
  const editor = createHeadlessEditor({
    namespace: 'protocol',
    nodes: nodes,
    onError: (error) => {
      console.error(error);
    }
  });

  const dummyId = "dummy-id";
  const dummyProvider: any = {
    awareness: {
      getLocalState: () => null,
      getStates: () => [],
    }
  }
  const copyTarget = new Doc();
  const copyBinding = createBinding(editor, dummyProvider, dummyId, copyTarget, new Map([[dummyId, copyTarget]]));

  const onYjsTreeChanges = (events: YEvent<YText>[], transaction: any) => {
    // Event target debug: event.target._collabNode: [CollabElementNode], event.target.toJSON() = ""
    syncYjsChangesToLexical(copyBinding, dummyProvider as any, events, false);
  };
  copyBinding.root.getSharedType().observeDeep(onYjsTreeChanges);

  // yDocState clearly has the bytes of the change (ie. length comparisons)
  applyUpdate(copyTarget, yDocState);
  
  editor.update(() => {}, { discrete: true });

  let markdown: string = "";
  editor.getEditorState().read(() => {
    markdown = $convertToMarkdownString(TRANSFORMERS);
  });

  // Here it is empty: markdown = ""
  // editor.getEditorState().toJSON() is also empty
  return markdown;
}
```

Loading the ydoc:
```typescript
const ydoc = new Doc();
const yjsDocMap: Map<string, Doc> = new Map([[id, ydoc]]);
const provider = createWebsocketProvider(id, yjsDocMap);
await provider.connect()
const yDocState = encodeStateAsUpdate(ydoc);
// yDocState.length => content loaded
```

**Reply by pinkevich** [reply]
Hello,
I faced the same problem with `_collabNode` on my websocket server, how did you solve it?
